objects/src/accounts/builder/mod.rs

```diff
@@ -0,0 +1,313 @@
+use alloc::vec::Vec;
+use core::fmt::Display;
+
+use crate::{
+    accounts::{Account, AccountComponent, AccountId, AccountStorageMode, AccountType},
+    assets::{Asset, AssetVault},
+    AccountError, AssetVaultError, Felt, Word, ZERO,
+};
+
+/// A convenient builder for an [`Account`] allowing for safe construction of an account by
+/// combining multiple [`AccountComponent`]s.
+///
+/// By default, the builder is initialized with:
+/// - The `nonce` set to [`ZERO`], i.e. the nonce of a new account.
+/// - The `account_type` set to [`AccountType::RegularAccountUpdatableCode`].
+/// - The `storage_mode` set to [`AccountStorageMode::Private`].
+///
+/// The methods that are required to be called are:
+///
+/// - [`AccountBuilder::init_seed`],
+/// - [`AccountBuilder::with_component`], which must be called at least once.
+#[derive(Debug, Clone)]
+pub struct AccountBuilder {
+    assets: Vec<Asset>,
+    components: Vec<AccountComponent>,
+    nonce: Felt,
+    account_type: AccountType,
+    storage_mode: AccountStorageMode,
+    init_seed: Option<[u8; 32]>,
+}
+
+impl AccountBuilder {
+    /// Creates a new builder for a single account.
+    pub fn new() -> Self {
+        Self {
+            assets: vec![],
+            components: vec![],
+            nonce: ZERO,
+            init_seed: None,
+            account_type: AccountType::RegularAccountUpdatableCode,
+            storage_mode: AccountStorageMode::Private,
+        }
+    }
+
+    /// Sets the initial seed from which the grind for an [`AccountId`] will start. This initial
+    /// seed should come from a cryptographic random number generator.
+    ///
+    ///  This method **must** be called.
+    pub fn init_seed(mut self, init_seed: [u8; 32]) -> Self {
+        self.init_seed = Some(init_seed);
+        self
+    }
+
+    /// Sets the type of the account.
+    pub fn account_type(mut self, account_type: AccountType) -> Self {
+        self.account_type = account_type;
+        self
+    }
+
+    /// Sets the storage mode of the account.
+    pub fn storage_mode(mut self, storage_mode: AccountStorageMode) -> Self {
+        self.storage_mode = storage_mode;
+        self
+    }
+
+    /// Sets the nonce of the account. This method is optional.
+    ///
+    /// If unset, the nonce will default to [`ZERO`].
+    pub fn nonce(mut self, nonce: Felt) -> Self {
+        self.nonce = nonce;
+        self
+    }
+
+    /// Adds the asset to the account's [`AssetVault`]. This method is optional.
+    pub fn with_asset(mut self, asset: Asset) -> Self {
+        self.assets.push(asset);
+        self
+    }
+
+    /// Adds all the assets to the account's [`AssetVault`]. This method is optional.
+    pub fn with_assets<I: IntoIterator<Item = Asset>>(mut self, assets: I) -> Self {
+        self.assets.extend(assets);
+        self
+    }
```

Comment:
We should probably mention in the doc-comments for these methods the condition I described in the above comment.